File: src/node/tui/rain.ts
// Matrix rain animation state machine
// Renders cursor-positioned falling characters in the rain zone

import { brightGreen, green, dimGreen } from "./colors.js";

// Character pools
const KATAKANA = "ｦｱｲｳｴｵｶｷｸｹｺｻｼｽｾｿﾀﾁﾂﾃﾄﾅﾆﾇﾈﾉﾊﾋﾌﾍﾎﾏﾐﾑﾒﾓﾔﾕﾖﾗﾘﾙﾚﾛﾜﾝ";
const DIGITS = "0123456789";
const LATIN = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
const CHAR_POOL = KATAKANA + DIGITS + LATIN;

const DENSITY = 0.3;
const DENSITY_REF_ROWS = 20; // zone height the current DENSITY was calibrated at
const MAX_DENSITY_SCALE = 3; // cap so very tall zones don't produce unbounded per-frame output
const MIN_SPEED = 0.3;
const MAX_SPEED = 1.0;
const MIN_LENGTH = 3;
const MAX_LENGTH = 8;
const MAX_RESPAWN_DELAY = 5;
const SHIMMER_RATE = 0.05;

function randFloat(min: number, max: number): number {
  return min + Math.random() * (max - min);
}

function randInt(min: number, max: number): number {
  return min + Math.floor(Math.random() * (max - min + 1));
}

function randChar(): string {
  return CHAR_POOL[Math.floor(Math.random() * CHAR_POOL.length)];
}

// Active drop count scales with rain-zone height: DENSITY is calibrated at
// DENSITY_REF_ROWS, and taller zones get proportionally more drops so coverage
// stays roughly constant. Clamped to 1x at/below the reference height
// (byte-identical to the prior width-only count) and capped at
// MAX_DENSITY_SCALE to bound per-frame output on very tall zones. With the
// shipped constants the count stays at or below `cols` (0.3 * 3 = 0.9); a
// column only hosts a second drop if DENSITY * MAX_DENSITY_SCALE is tuned
// above 1.
export function computeActiveDropCount(cols: number, rows: number): number {
  const heightScale = Math.min(MAX_DENSITY_SCALE, Math.max(1, rows / DENSITY_REF_ROWS));
  return Math.round(cols * DENSITY * heightScale);
}

interface Drop {
  col: number;
  row: number;       // fractional position (e.g., 5.6)
  speed: number;     // fractional rows per tick (0.3-1.0)
  length: number;    // trail length
  delay: number;     // ticks before starting
  chars: string[];   // pre-generated characters
}

export class RainState {
  private cols: number;
  private rows: number;
  private startCol: number;  // 0-based column offset for right-margin mode
  private drops: Drop[];
  private prevPositions: Set<string>; // "row,col" keys for previous frame

  constructor(cols: number, rows: number, startCol = 0) {
    this.cols = cols;
    this.rows = rows;
    this.startCol = startCol;
    this.drops = [];
    this.prevPositions = new Set();
    this.initDrops();
  }

  private initDrops(): void {
    this.drops = [];
    this.prevPositions = new Set();
    const activeCount = computeActiveDropCount(this.cols, this.rows);
    // Shuffle column indices, then assign drops round-robin over them: every
    // column is used once before any column would get a second drop (which
    // only happens if the constants are tuned so DENSITY * MAX_DENSITY_SCALE
    // exceeds 1 — shipped constants stay within one drop per column).
    const columns = Array.from({ length: this.cols }, (_, i) => i);
    for (let i = columns.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [columns[i], columns[j]] = [columns[j], columns[i]];
    }
    for (let i = 0; i < activeCount; i++) {
      this.drops.push(this.createDrop(columns[i % this.cols], true));
    }
  }

  private createDrop(col: number, scatter: boolean): Drop {
    const length = randInt(MIN_LENGTH, MAX_LENGTH);
    const chars = Array.from({ length }, () => randChar());
    return {
      col,
      row: scatter ? randFloat(-length, this.rows) : -length,
      speed: randFloat(MIN_SPEED, MAX_SPEED),
      length,
      delay: scatter ? 0 : randInt(0, MAX_RESPAWN_DELAY),
      chars,
    };
  }

  tick(): void {
    for (const drop of this.drops) {
      if (drop.delay > 0) {
        drop.delay--;
        continue;
      }
      drop.row += drop.speed;

      // Shimmer: randomly replace ~5% of trail chars
      for (let i = 0; i < drop.chars.length; i++) {
        if (Math.random() < SHIMMER_RATE) drop.chars[i] = randChar();
      }

      // Respawn when fully off-screen
      if (Math.floor(drop.row) - drop.length > this.rows) {
        const newDrop = this.createDrop(drop.col, false);
        drop.row = newDrop.row;
        drop.speed = newDrop.speed;
        drop.length = newDrop.length;
        drop.delay = newDrop.delay;
        drop.chars = newDrop.chars;
      }
    }
  }

  render(startRow: number): string {
    if (this.rows <= 0) return "";

    let output = "";
    const currentPositions = new Set<string>();

    for (const drop of this.drops) {
      if (drop.delay > 0) continue;

      const headRow = Math.floor(drop.row);

      for (let i = 0; i < drop.length; i++) {
        const row = headRow - i;
        if (row < 0 || row >= this.rows) continue;

        const screenRow = startRow + row;
        const screenCol = this.startCol + drop.col + 1; // 1-based for ANSI
        const char = drop.chars[i % drop.chars.length];

        const key = `${row},${drop.col}`;
        currentPositions.add(key);

        let colored: string;
        if (i === 0) {
          colored = brightGreen(char);
        } else if (i < drop.length - 2) {
          colored = green(char);
        } else {
          colored = dimGreen(char);
        }

        output += `\x1b[${screenRow};${screenCol}H${colored}`;
      }
    }

    // Clear old positions that are no longer occupied
    for (const key of this.prevPositions) {
      if (!currentPositions.has(key)) {
        const [r, c] = key.split(",").map(Number);
        const screenRow = startRow + r;
        const screenCol = this.startCol + c + 1; // 1-based for ANSI
        if (r >= 0 && r < this.rows) {
          output += `\x1b[${screenRow};${screenCol}H `;
        }
      }
    }

    this.prevPositions = currentPositions;

    return output;
  }

  resize(cols: number, rows: number, startCol = 0): void {
    if (cols === this.cols && rows === this.rows && startCol === this.startCol) return; // no-op
    this.cols = cols;
    this.rows = rows;
    this.startCol = startCol;
    this.initDrops();
  }
}

File: src/go/internal/watch/rain.go
package watch

import (
	"math"
	"math/rand/v2"
	"strings"

	"github.com/sahil87/tu/internal/render"
	"github.com/sahil87/tu/internal/render/ansi"
)

// The rain character pool (the TS rain.ts): half-width katakana + digits +
// A–Za–z. Every character is BMP, so rune indexing matches the TS code-unit
// indexing.
var rainPool = []rune("ｦｱｲｳｴｵｶｷｸｹｺｻｼｽｾｿﾀﾁﾂﾃﾄﾅﾆﾇﾈﾉﾊﾋﾌﾍﾎﾏﾐﾑﾒﾓﾔﾕﾖﾗﾘﾙﾚﾛﾜﾝ" +
	"0123456789" +
	"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz")

// The rain tuning constants (the TS rain.ts): the density calibrated at 20
// rows, the 3× scale cap, drop speed 0.3–1.0 rows/tick, trail length 3–8,
// respawn delay 0–5 ticks, shimmer rate 0.05.
const (
	rainDensity         = 0.3
	rainDensityRefRows  = 20
	rainMaxDensityScale = 3
	rainMinSpeed        = 0.3
	rainMaxSpeed        = 1.0
	rainMinLength       = 3
	rainMaxLength       = 8
	rainMaxRespawnDelay = 5
	rainShimmerRate     = 0.05
)

// ActiveDropCount is the TS computeActiveDropCount: the count scales with the
// zone HEIGHT (density × heightScale, clamped 1× at/below 20 rows, capped 3×
// at 60+), rounded with the JS Math.round rule.
func ActiveDropCount(cols, rows int) int {
	heightScale := math.Min(rainMaxDensityScale, math.Max(1, float64(rows)/rainDensityRefRows))
	return int(render.JSRound(float64(cols) * rainDensity * heightScale))
}

// rainDrop is one falling character trail (the TS Drop).
type rainDrop struct {
	col    int
	row    float64 // fractional position
	speed  float64 // rows per tick
	length int     // trail length
	delay  int     // ticks before starting
	chars  []rune  // pre-generated
}

// RainState is the rain layer's state machine (the TS RainState) with an
// injected RNG and Colors. The previous frame's occupied cells are kept in
// INSERTION order (the JS Set's iteration order) so the vacated-cell clears
// emit in a deterministic order.
type RainState struct {
	cols, rows, startCol int // startCol is the 0-based column offset (right-margin mode)
	rng                  *rand.Rand
	colors               ansi.Colors
	drops                []rainDrop
	prev                 [][2]int // occupied (row, col) of the previous render, insertion order
}

// NewRainState builds the state and scatters the initial drops (the TS
// constructor + initDrops): a Fisher–Yates shuffle of the column indices,
// then drop i on columns[i % cols] (round-robin; the shipped constants never
// exceed one drop per column).
func NewRainState(cols, rows, startCol int, rng *rand.Rand, c ansi.Colors) *RainState {
	r := &RainState{cols: cols, rows: rows, startCol: startCol, rng: rng, colors: c}
	r.initDrops()
	return r
}

// Resize keeps the drops when (cols, rows, startCol) are unchanged and
// re-initialises otherwise (the TS resize; startRow is a render-time input,
// not part of the geometry).
func (r *RainState) Resize(cols, rows, startCol int) {
	if cols == r.cols && rows == r.rows && startCol == r.startCol {
		return
	}
	r.cols, r.rows, r.startCol = cols, rows, startCol
	r.initDrops()
}

func (r *RainState) initDrops() {
	r.drops = nil
	r.prev = nil
	count := ActiveDropCount(r.cols, r.rows)
	columns := make([]int, r.cols)
	for i := range columns {
		columns[i] = i
	}
	for i := len(columns) - 1; i > 0; i-- {
		j := int(r.rng.Float64() * float64(i+1))
		columns[i], columns[j] = columns[j], columns[i]
	}
	for i := 0; i < count; i++ {
		r.drops = append(r.drops, r.createDrop(columns[i%r.cols], true))
	}
}

// createDrop is the TS createDrop: the RNG draw order is length, the chars
// (length draws), the scattered row (scatter only), speed, and the respawn
// delay (respawn only) — the seeded golden depends on it.
func (r *RainState) createDrop(col int, scatter bool) rainDrop {
	length := r.randInt(rainMinLength, rainMaxLength)
	chars := make([]rune, length)
	for i := range chars {
		chars[i] = r.randChar()
	}
	d := rainDrop{col: col, length: length, chars: chars}
	if scatter {
		d.row = r.randFloat(-float64(length), float64(r.rows))
	} else {
		d.row = -float64(length)
	}
	d.speed = r.randFloat(rainMinSpeed, rainMaxSpeed)
	if !scatter {
		d.delay = r.randInt(0, rainMaxRespawnDelay)
	}
	return d
}

// Tick advances every drop (the TS tick): a delayed drop counts down; others
// move by their speed, shimmer (each trail char replaced with probability
// 0.05), and respawn in place when fully off-screen (row − length > rows:
// back to −length with new speed, length, delay and chars).
func (r *RainState) Tick() {
	for i := range r.drops {
		d := &r.drops[i]
		if d.delay > 0 {
			d.delay--
			continue
		}
		d.row += d.speed
		for j := range d.chars {
			if r.rng.Float64() < rainShimmerRate {
				d.chars[j] = r.randChar()
			}
		}
		if math.Floor(d.row)-float64(d.length) > float64(r.rows) {
			nd := r.createDrop(d.col, false)
			d.row, d.speed, d.length, d.delay, d.chars = nd.row, nd.speed, nd.length, nd.delay, nd.chars
		}
	}
}

// Render emits the cursor-positioned frame (the TS render): for each active
// drop and trail index, `\x1b[{startRow+r};{startCol+col+1}H` + the colored
// char (BrightGreen head, Green body, DimGreen the last two), then a
// `\x1b[{r};{c}H ` clear for every previously occupied cell not occupied now.
// "" when rows <= 0.
func (r *RainState) Render(startRow int) string {
	if r.rows <= 0 {
		return ""
	}
	var b strings.Builder
	cur := make([][2]int, 0, len(r.prev))
	curSet := make(map[[2]int]bool, len(r.prev))

	for _, d := range r.drops {
		if d.delay > 0 {
			continue
		}
		head := int(math.Floor(d.row))
		for i := 0; i < d.length; i++ {
			row := head - i
			if row < 0 || row >= r.rows {
				continue
			}
			key := [2]int{row, d.col}
			if !curSet[key] {
				curSet[key] = true
				cur = append(cur, key)
			}
			ch := d.chars[i%d.length]
			var colored string
			switch {
			case i == 0:
				colored = r.colors.BrightGreen(string(ch))
			case i < d.length-2:
				colored = r.colors.Green(string(ch))
			default:
				colored = r.colors.DimGreen(string(ch))
			}
			b.WriteString("\x1b[" + itoa(startRow+row) + ";" + itoa(r.startCol+d.col+1) + "H" + colored)
		}
	}

	for _, key := range r.prev {
		if !curSet[key] {
			row, col := key[0], key[1]
			if row >= 0 && row < r.rows {
				b.WriteString("\x1b[" + itoa(startRow+row) + ";" + itoa(r.startCol+col+1) + "H ")
			}
		}
	}

	r.prev = cur
	return b.String()
}

// randFloat is the TS randFloat: min + random() × (max − min).
func (r *RainState) randFloat(minV, maxV float64) float64 {
	return minV + r.rng.Float64()*(maxV-minV)
}

// randInt is the TS randInt: min + floor(random() × (max − min + 1)).
func (r *RainState) randInt(minV, maxV int) int {
	return minV + int(r.rng.Float64()*float64(maxV-minV+1))
}

// randChar is the TS randChar: a uniform pool character.
func (r *RainState) randChar() rune {
	return rainPool[int(r.rng.Float64()*float64(len(rainPool)))]
}
